src/integrationTest/sam.test.ts

```diff
@@ -0,0 +1,181 @@
+import { mkdirpSync, readFileSync, removeSync } from 'fs-extra'
+import * as path from 'path'
+import * as vscode from 'vscode'
+import { SamLambdaRuntime } from '../../src/lambda/models/samLambdaRuntime'
+import { getSamCliContext } from '../../src/shared/sam/cli/samCliContext'
+import { runSamCliInit, SamCliInitArgs } from '../../src/shared/sam/cli/samCliInit'
+import { activateExtension, sleep, TIMEOUT } from './integrationTestsUtilities'
+
+const projectFolder = `${__dirname}`
+let projectSDK = ''
+let projectPath = ''
+let debuggerType = ''
+let documentUri: vscode.Uri
+
+const runtimes = [
+    { name: 'nodejs10.x', path: 'testProject/hello-world/app.js', debuggerType: 'node2' },
+    { name: 'python2.7', path: 'testProject/hello_world/app.py', debuggerType: 'python' },
+    { name: 'python3.6', path: 'testProject/hello_world/app.py', debuggerType: 'python' },
+    { name: 'python3.7', path: 'testProject/hello_world/app.py', debuggerType: 'python' },
+    { name: 'dotnetcore2.1', path: 'testProject/src/HelloWorld/Function.cs', debuggerType: 'coreclr' }
+]
+
+async function openSamProject(): Promise<vscode.Uri> {
+    const documentPath = path.join(projectFolder, projectPath)
+    await vscode.workspace.openTextDocument(documentPath)
+
+    return vscode.Uri.file(documentPath)
+}
+
+async function getCodeLenses(): Promise<vscode.CodeLens[]> {
+    let codeLenses: vscode.CodeLens[] | undefined
+    while (true) {
+        try {
+            const codeLensesPromise: Thenable<vscode.CodeLens[] | undefined> =
+                vscode.commands.executeCommand('vscode.executeCodeLensProvider', documentUri)
+            codeLenses = await codeLensesPromise
+            if (!codeLenses) {
+                continue
+            }
+            // omnisharp spits out some undefined code lenses for some reason, we filter them because they are
+            // not shown to the user and do not affect how our extension is working
+            codeLenses = codeLenses!.filter(lens => lens !== undefined && lens.command !== undefined)
+            if (codeLenses!.length === 3) {
+                return codeLenses as vscode.CodeLens[]
+            }
+        } catch (e) { }
+    }
+}
+
+async function getCodeLensesOrTimeout(): Promise<vscode.CodeLens[]> {
+    const codeLensPromise = getCodeLenses()
+    const timeout = new Promise((resolve) => { setTimeout(resolve, 10000, undefined) })
+    const result = await Promise.race([codeLensPromise, timeout])
+
+    if (result) {
+        return result as vscode.CodeLens[]
+    }
+    throw new Error('Codelenses took too long to show up, this inidicates an issue!')
+}
+
+async function onDebugChanged(e: vscode.DebugSession | undefined) {
+    if (!e) {
+        return
+    }
+    assert.strictEqual(e.configuration.name, 'SamLocalDebug')
+    assert.strictEqual(e.configuration.type, debuggerType)
+    // wait for it to actually start (which we do not get an event for)
+    await sleep(800)
```

Comment:
Is 800 used for a specific reason? Should we look to emit an event instead?

Reply:
There is no event emitted, I'll append why 800